### ISSUE TITLE
[Kobo] Charging led does not light up

### ISSUE THREAD

**Issue by SilverGreen93** [koreader/koreader#6527]
* KOReader version: 2020.08
* Device: Kobo Forma

#### Issue
When plugging in the usb charger, the charging led does not light up even when the device is sleeping, so there is no indication of the chraging state.

#### Steps to reproduce
Open KOReader.
Put the device to sleep.
Plug in the USB charger.
No indication of charging is shown.
Charging LED does not light up.

**Comment by pazos** [comment]
IMO too much work for litle benefit. Patches welcome

**Comment by NiLuJe** [comment]
We've already established that playing with the LEDs led to... unwanted and weird behavior in weird and interesting ways, so, yeaaah, not happening ;).

**Comment by NiLuJe** [comment]
(I can't *quite* recall the issue where we played with this last, but it was a couple years ago).

**Comment by SilverGreen93** [comment]
Maybe at least display on the lockscreen the charging message and update it accordingly?

**Comment by NiLuJe** [comment]
I would think the giant cable hanging from the device would be obvious enough?

(We don't update the screen on suspend, so having a live tracking % is not happening either).

**Comment by Frenzie** [comment]
That means it either doesn't actually sleep while charging or that it wakes up every x minutes to update the status.

**Comment by NiLuJe** [comment]
And that's potentially device-specific. Mk. 7 might be able to suspend while charging, like they do when USBNet is enabled.

While earlier devices will actually fail to sleep.

(FWIW, Nickel itself prevents suspend entirely when charging, on every generation).

**Comment by SilverGreen93** [comment]
> I would think the giant cable hanging from the device would be obvious enough?
> 
> (We don't/can't update the screen on suspend, so having a live tracking % is not happening either).

No, the cable connected to the device does not mean it's charging...
On my first Kobo device I had a defective microUSB connector and didn't know that it wasn't charging, so I replaced the defective unit. It was charging only if the connector was held at a specific angle. Cables can get broken etc...

Yes, Nickel wakes during charging and updates the screen.

**Comment by NiLuJe** [comment]
> No, the cable connected to the device does not mean it's charging...
> On my first Kobo device I had a defective microUSB connector and didn't know that it wasn't charging, so I replaced the defective unit. It was charging only if the connector was held at a specific angle. Cables can get broken etc...
> 

That's veering into extreme edge-casey behavior ^^.

**Comment by NiLuJe** [comment]
Just double-checked, for shit'n giggles, and, yeah, a Mk. 7 device will happily successfully suspend *while charging*.

Score on for smarter devices!

----

In any case, I have minus 100 interest points in looking into this, but PRs welcome ;).

A few pointers:

* There's a token system in the suspend message widget, IIRC, so adding one for battery % if that's not already in there should be trivial.
* The suspend code is... extremely fragile. So expect *stringent* reaction against touching it at all on Kobo.
* powerd should have a pretty accurate idea of whether we're charging or not. But you might need to resort to the dedicated ntx_io ioctl to check if it's plugged in vs. charging. or something. Been a while since I checked, can't recall how accurate sysfs is compared to the ioctls.

**Comment by Frenzie** [comment]
@NiLuJe Hm, I thought my H2O v1 did too? But I might be misremembering.

**Comment by NiLuJe** [comment]
> @NiLuJe Hm, I thought my H2O v1 did too? But I might be misremembering.

I know it won't in USBNet.

I would assume that meant it wouldn't when just charging, but I don't recall having double-checked recently. A bit easier to do now that I have a klogd in my bb applet, so I get timestamped kernel logs in the syslog, though.

**Comment by Frenzie** [comment]
> I know it won't in USBNet.

That it won't for sure. Pretty different imho. :-)
